src/Servers/Kestrel/shared/test/TransportTestHelpers/TestServer.cs

```diff
@@ -70,32 +70,37 @@ public TestServer(RequestDelegate app, TestServiceContext context, Action<Kestre
-                })
-                .ConfigureServices(services =>
-                {
-                    services.AddSingleton<IStartup>(this);
-                    services.AddSingleton(context.LoggerFactory);
-                    services.AddSingleton<IServer>(sp =>
-                    {
-                        // Manually configure options on the TestServiceContext.
-                        // We're doing this so we can use the same instance that was passed in
-                        var configureOptions = sp.GetServices<IConfigureOptions<KestrelServerOptions>>();
-                        foreach (var c in configureOptions)
+                    webHostBuilder
+                        .UseKestrel(options =>
                         {
-                            c.Configure(context.ServerOptions);
-                        }
-
-                        return new KestrelServer(new List<IConnectionListenerFactory>() { sp.GetRequiredService<IConnectionListenerFactory>() }, context);
-                    });
-                    configureServices(services);
+                            configureKestrel(options);
+                            _listenOptions = options.ListenOptions.First();
+                        })
+                        .ConfigureServices(services =>
+                        {
+                            services.AddSingleton<IStartup>(this);
+                            services.AddSingleton(context.LoggerFactory);
+                            services.AddSingleton<IServer>(sp =>
+                            {
+                                // Manually configure options on the TestServiceContext.
+                                // We're doing this so we can use the same instance that was passed in
+                                var configureOptions = sp.GetServices<IConfigureOptions<KestrelServerOptions>>();
+                                foreach (var c in configureOptions)
+                                {
+                                    c.Configure(context.ServerOptions);
+                                }
+
+                                return new KestrelServer(new List<IConnectionListenerFactory>() { sp.GetRequiredService<IConnectionListenerFactory>() }, context);
+                            });
+                            configureServices(services);
+                        })
+                        .UseSetting(WebHostDefaults.ApplicationKey, typeof(TestServer).GetTypeInfo().Assembly.FullName)
+                        .UseSetting(WebHostDefaults.ShutdownTimeoutKey, TestConstants.DefaultTimeout.TotalSeconds.ToString())
+                        .Configure(app => { app.Run(_app); });
                 })
-                .UseSetting(WebHostDefaults.ApplicationKey, typeof(TestServer).GetTypeInfo().Assembly.FullName)
-                .UseSetting(WebHostDefaults.ShutdownTimeoutKey, TestConstants.DefaultTimeout.TotalSeconds.ToString())
                 .Build();
 
             _host.Start();
```

Comment:
It doesn't have to be done for this PR, but long term I'd like to get rid of all calls to the IHost.Start() extension method and calls IHost.Dispose() in our tests and instead use their async alternatives. This will mean moving away from using the TestServer constructor to give as an already-started TestServer. We might need to switch to an async TestServer factory method instead.